### PR TITLE
SPDK get NVMe SMART page and NVMe Intel read/write latency pages.

### PR DESCRIPTION
**Pull request by bsedyang**
Add spdk_bdev_get_smart(), spdk_bdev_set_latency_tracking(), spdk_bdev_get_latency() at bdev layer and the related blockdev_nvme functions for retrieving those pages. As for the latency page, the average latency is calculated and put where the Major revision and Minor revision were.

### CODE REVIEW

File: lib/bdev/nvme/blockdev_nvme.c [hunk 1]
```diff
@@ -311,6 +391,8 @@ blockdev_nvme_create_cb(void *io_device, uint32_t priority, void *ctx_buf, void
 
 	spdk_poller_register(&ch->poller, blockdev_nvme_poll, ch->qpair,
 			     spdk_app_get_current_core(), NULL, 0);
+	spdk_poller_register(&ch->adminq_poller, blockdev_nvme_poll_adminq, ctrlr,
```
Comment (dverkamp-intel) [inline lib/bdev/nvme/blockdev_nvme.c]: This will cause an admin queue poller to be registered for each I/O channel, but there is only one admin queue per NVMe controller. This will lead to poor performance due to the locking required, and I don't think the current implementation in this patch will behave correctly when used with multiple I/O channels per controller.

I believe it would be better to have a single admin queue poller per NVMe controller that should process all admin completions and use the SPDK event library function spdk_event_call() to pass the completion back to the correct thread.  This admin poller could be registered in attach_cb().

File: lib/nvme/nvme_ctrlr_cmd.c [hunk 2]
```diff
@@ -438,3 +438,18 @@ nvme_ctrlr_cmd_fw_image_download(struct spdk_nvme_ctrlr *ctrlr,
 
 	return rc;
 }
+
+inline int32_t
```
Comment (dverkamp-intel) [inline lib/nvme/nvme_ctrlr_cmd.c]: These new APIs should be unnecessary if the NVMe bdev code tracks its own outstanding admin commands (see comment on blockdev_nvme.c).

**Comment by pvonstamwitz** [comment]
We understand your concern exposing vendor-specific commands through the bdev layer. Therefore, we will take your advice and split this PR. We will resubmit with SPDK_BDEV_IO_TYPE_SMART_READ_DATA only with your suggested changes.

As for latency tracking information, our use case is a monitoring service that will allow for a user to measure latency over a user-defined time frame for PCIe-NVME devices. The same monitoring service will be periodically polling SPDK_BDEV_IO_TYPE_SMART_READ_DATA. The latency commands from the NVMe device is very useful for this, and we'd like to explore ways this can be supported without having the monitoring service call directly to the PMD. Should we consider some sort of plug-in model?

**Comment by dverkamp-intel** [comment]
Okay, I think we can figure out a reasonable way to expose latency information, but we probably need to do more than just expose the raw vendor-specific commands.

One issue is that the Set Features for enabling and disabling latency tracking acts on the whole NVMe controller, but a bdev represents a single NVMe namespace.  This is not an issue for the Intel SSD  DC P3700 family, since these devices only support a single namespace, but other devices may export multiple namespaces, and toggling the latency tracking feature via the bdev interface would affect other bdevs (namespaces on the same controller).

**Comment by bsedyang** [comment]
Code for the latency commands was removed.
One admin poller per controller now is registered in attach_cb.
A SMART specific union is added into struct spdk_bdev_io.  To avoid unnecessary polling, a new member, outstanding_admin_cmds, is added into struct nvme_device.

File: lib/bdev/nvme/blockdev_nvme.c [hunk 3]
```diff
@@ -673,6 +700,29 @@ queued_next_sge(void *ref, void **address, uint32_t *length)
 	return 0;
 }
 
+static void
+queued_done_adminq(void *ref, const struct spdk_nvme_cpl *cpl)
```
Comment (dverkamp-intel) [inline lib/bdev/nvme/blockdev_nvme.c]: This function looks like it mostly does the same as queued_done, so it would be nice to just decrement outstanding_admin_cmds and then call queued_done to reuse the common code.

File: lib/bdev/nvme/blockdev_nvme.c [hunk 4]
```diff
@@ -746,6 +796,21 @@ blockdev_nvme_unmap(struct nvme_blockdev *nbdev, struct spdk_io_channel *ch,
...
+static int
+blockdev_nvme_smart_read_data(struct nvme_blockdev *nbdev, struct nvme_blockio *bio, void *buf)
+{
+	if (spdk_nvme_ctrlr_cmd_get_log_page(nbdev->ctrlr, SPDK_NVME_LOG_HEALTH_INFORMATION,
+					     SPDK_NVME_GLOBAL_NS_TAG, buf, 512,
+					     queued_done_adminq, bio)) {
+		return -1;
+	} else {
+		/* increment the outstanding admin commands by 1 */
+		nbdev->nvme_dev->outstanding_admin_cmds++;
```
Comment (dverkamp-intel) [inline lib/bdev/nvme/blockdev_nvme.c]: This is modifying global data without any synchronization.  The simplest option to fix this would be to turn this into an atomic increment (__sync_fetch_and_add() or similar) and also change the corresponding decrement.

File: lib/bdev/nvme/blockdev_nvme.c [hunk 5]
```diff
@@ -231,6 +246,7 @@ static void blockdev_nvme_get_rbuf_cb(struct spdk_bdev_io *bdev_io)
 
 static int _blockdev_nvme_submit_request(struct spdk_bdev_io *bdev_io)
 {
+
```
Comment (dverkamp-intel) [inline lib/bdev/nvme/blockdev_nvme.c]: Unrelated whitespace change

File: lib/bdev/nvme/blockdev_nvme.c [hunk 6]
```diff
@@ -262,6 +278,11 @@ static int _blockdev_nvme_submit_request(struct spdk_bdev_io *bdev_io)
 					   bdev_io->u.flush.offset,
 					   bdev_io->u.flush.length);
 
+	case SPDK_BDEV_IO_TYPE_SMART:
```
Comment (dverkamp-intel) [inline lib/bdev/nvme/blockdev_nvme.c]: The new SPDK_IO_TYPE should also be added to blockdev_nvme_io_type_supported().

File: lib/bdev/nvme/blockdev_nvme.c (hunk 3 above)
Comment (bsedyang) [inline lib/bdev/nvme/blockdev_nvme.c]: which means the spdk_bdev_io_from_ctx() will be called twice, one in queued_done_adminq and the other one in queued_done, am I right?

**Comment by benlwalker** [comment]
To address the piping of the latency command through the system, what if we define a new bdev command that is an NVME_PASSTHROUGH command? This would allow raw NVMe commands to be passed through the bdev layer. Virtual bdevs and non-NVMe bdevs could just fail all NVME_PASSTHROUGH commands.

**Comment by pvonstamwitz** [comment]
Hi Ben, we thought of the pass-through as well as direct mode. But, what about providing a generic latency command which provides a simple format of number of reads/writes, average latency, and a mechanism to toggle measurement? The blockdev_nvme layer can convert it to the controller specific commands and convert the output to the generic format. Other device types could use different mechanisms to provide the data or just fail the command.

**Comment by pvonstamwitz** [comment]
Can you please review the latest commit 4b6187d?

**Comment by bsedyang** [comment]
Hi Daniel,

I saw the conflict against the spdk/master.  If there is anything I should do to resolve the conflict, please let me know.  Thanks.

**Comment by dverkamp-intel** [comment]
It would be helpful if you could upload a version that is rebased on the latest master (or has the latest master branch merged into it).

I don’t think there should be any major conflicts – if you run into any problems, please let us know.

**Comment by bsedyang** [comment]
Hi Daniel,

The conflict is resolved.
Would you please review it again?

Thanks,
Edward

**Comment by bsedyang** [comment]
Hi Daniel,

If there is anything we can (or should) do in speed-up of this pull request, please let us know.

Thank you very much,
Edward

File: include/spdk/bdev.h [hunk 7]
```diff
@@ -145,6 +146,56 @@ enum spdk_bdev_reset_type {
...
+		uint8_t	volatile_memory_backup	: 1;
+		uint8_t	reserved		: 3;
+	} bits;
+};
+SPDK_STATIC_ASSERT(sizeof(union spdk_critical_warning_state) == 1, "Incorrect size");
+
+/**
+ * health report page
+ */
+struct __attribute__((packed)) spdk_health_report_page {
```
Comment (dverkamp-intel) [inline include/spdk/bdev.h]: This looks like it is just a copy of the NVMe health data - if that is the intent, it should just use the structures from nvme_spec.h directly.

**Comment by pvonstamwitz** [comment]
Currently, all we need is the NVME health data, but our intention is to allow this structure to be extensible to provide information for non-NMVe devices in the future.

**Comment by bsedyang** [comment]
Hi Daniel,

The conflict was resolved due to removing spdk_bdev_io_submit() from API in bdev.h
Would you please review the changes?

Thanks,
Edward

**Comment by pvonstamwitz** [comment]
Please let us know what we need to do to get this merged. This PR enables polling for admin queues and we need this for some for some other functionality, i.e. abort commands on timeouts.

**Comment by dverkamp-intel** [comment]
Sorry for the slow review process on this patch, and thanks for your patience.

Since you are looking for the admin queue functionality in the bdev, we can split that part of the change out to a separate patch; I don't think that will be controversial.  We want to get the bdev API right and avoid breaking changes/maintenance burden later on, but just adding the admin queue poller will not affect bdev public API.

I will split the patch up and review/merge just the admin queue poller enablement, if you are okay with that.  Then we can have more discussion about the best way to represent the health info you need via the bdev API.

**Comment by pvonstamwitz** [comment]
Yes, splitting the patch up and merging just the admin queue poller enablement for now would be great.
Thanks!

**Comment by bsedyang** [comment]
If I understood correctly, you would review/merge the admin queue poller enablement portion.
Any help needed, please let me know.  Thanks.

**Comment by bsedyang** [comment]
Hi Daniel,

I know you are busy; hence, I wonder if it is OK for me to just close this pull request and re-submit a new one with just admin queue poller enabling.  Thanks.

**Comment by dverkamp-intel** [comment]
I have checked in an initial implementation of an admin queue poller for the NVMe bdev in commit 5de35015b92acfe30f9debaa6ffd74a410b35d74.  Note that this commit only sets up a timer-based poller, which can add potentially unacceptable amounts of latency if this poller were to be used for command completions, but we can use it as a starting point.

I think the best approach for bdev requests that need to issue admin commands is to temporarily register a second (non-timer-based) poller during the time that an admin command is outstanding.  This poller should  run on the same lcore as the timer-based poller to avoid thread safety issues.  However, before implementing this complexity, I think it would be best to use the timer-based poller for the initial implementation, then add the second as-needed poller once that is working.

Implementing this correctly will probably require the use of the event framework to pass a message from the admin queue poller thread back to the original bdev request processing thread to signal completion of the admin command.  This is because the admin queue poller may not be running on the same lcore as the bdev request processing thread.  An event that completes the bdev request can be created via spdk_event_allocate() in the bdev request thread and passed as the context to the NVMe admin command completion callback, and the NVMe completion callback should send the event with spdk_event_call().

**Comment by bsedyang** [comment]
Thanks, Daniel. 
Hence, I am closing this pull request.
